### PR TITLE
[plug-in][languages] Register a references provider #3308

### PR DESCRIPTION
**Pull request by vrubezhny**
Issue #3308

Signed-off-by: Victor Rubezhny <vrubezhny@redhat.com>

### CODE REVIEW

File: packages/plugin/API.md [hunk 1]
```diff
@@ -519,6 +519,27 @@ function provideDefinitionHandler(document: theia.TextDocument, position: theia.
...
+
+```typescript
+const documentsSelector: theia.DocumentSelector = { scheme: 'file', language: 'typescript' };
+const handler: theia.ReferenceProvider = { provideReferences: provideReferencesHandler };
+
+const disposable = theia.languages.registerReferenceProvider(documentsSelector, handler);
+
+...
+
+function provideRefereencesHandler(document: theia.TextDocument, position: theia.Position, context: theia.ReferenceContext): theia.ProviderResult<theia.Location[]> {
```
Comment (benoitf) [inline packages/plugin/API.md]: it seems that there is a typo on provideRefereencesHandler (two e)

File: packages/plugin-ext/src/plugin/languages/reference.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                }
+
+                return locations;
+            }
+        });
+    }
+}
+
+/* tslint:disable-next-line:no-any */
+function isLocationArray(array: any): array is types.Location[] {
```
Comment (benoitf) [inline packages/plugin-ext/src/plugin/languages/reference.ts]: why is it not a private method ? also do we have type of reference so we could avoid any ?

File: packages/plugin/API.md (hunk 1 above)
Comment (vrubezhny) [reply packages/plugin/API.md]: Yes, there is. I'll provide a fixup

File: packages/plugin-ext/src/plugin/languages/reference.ts (hunk 2 above)
Comment (vrubezhny) [reply packages/plugin-ext/src/plugin/languages/reference.ts]: `isLocationArray` is not exported, so it's visible only inside its module, I can't put `private` modifier on it (unless I make it a member of `ReferenceAdapter` class). Honestly, it's `copy-pasted` from `./definition.ts`, but I think, as it's simple and duplicated only once,  it's reasonable to leave it as it is, or, if needed later, to move such functions into some `utility.ts`-like module and make them exported.

According to https://github.com/Microsoft/vscode/blob/5a4eab812bd485942061f71af39f3eabc52a1035/src/vs/vscode.d.ts#L2604 i don't see any kind of `types` for the references.

Comment (benoitf) [reply packages/plugin-ext/src/plugin/languages/reference.ts]: if it's only used internally, then I still don't see why not use it as private method of ReferenceAdapter ?

**Comment by benoitf** [comment]
FYI merge commits can't be there in theia (only rebase option is available to merge PR)

**Comment by vrubezhny** [comment]
OK, I'll rebase my branch then.

File: packages/plugin-ext/src/plugin/languages/reference.ts (hunk 2 above)
Comment (vrubezhny) [reply packages/plugin-ext/src/plugin/languages/reference.ts]: I've added a fixup for this.

**Comment by benoitf** [comment]
@vrubezhny could you squash the 3 commits into 1 (as only rebase and merge option is available)

**Pull request merged**